File: 3558. Number of Ways to Assign Edge Weights I.py
class Solution:
    def assignEdgeWeights(
        self,
        edge_list: List[List[int]]
    ) -> int:

        MODULO = 10**9 + 7

        total_nodes = len(edge_list) + 1

        # Adjacency list representation of the tree
        adjacency_list = [
            [] for _ in range(total_nodes + 1)
        ]

        # Build the graph
        for source_node, destination_node in edge_list:

            adjacency_list[source_node].append(
                destination_node
            )

            adjacency_list[destination_node].append(
                source_node
            )

        # Maximum depth found in the tree
        maximum_depth = 0

        # (node, depth)
        dfs_stack = [(1, 0)]

        visited_nodes = [False] * (
            total_nodes + 1
        )

        visited_nodes[1] = True

        # Iterative DFS
        while dfs_stack:

            current_node, current_depth = (
                dfs_stack.pop()
            )

            maximum_depth = max(
                maximum_depth,
                current_depth
            )

            for neighbor_node in adjacency_list[current_node]:

                if not visited_nodes[neighbor_node]:

                    visited_nodes[neighbor_node] = True

                    dfs_stack.append(
                        (
                            neighbor_node,
                            current_depth + 1
                        )
                    )

        # Return 2^(maximum_depth - 1) mod MODULO
        return pow(
            2,
            maximum_depth - 1,
            MODULO
        )

File: 3558. Number of Ways to Assign Edge Weights I.cpp
class Solution {
public:

    int assignEdgeWeights(vector<vector<int>>& edgeList) {

        const int MODULO = 1e9 + 7;

        int totalNodes = edgeList.size() + 1;

        // Adjacency list representation of the tree
        vector<vector<int>> adjacencyList(
            totalNodes + 1
        );

        // Build the graph
        for (const auto& edge : edgeList) {

            int sourceNode = edge[0];
            int destinationNode = edge[1];

            adjacencyList[sourceNode].push_back(
                destinationNode
            );

            adjacencyList[destinationNode].push_back(
                sourceNode
            );
        }

        // Stores the maximum depth of the tree
        int maximumDepth = 0;

        // {node, depth}
        stack<pair<int, int>> dfsStack;

        dfsStack.push({1, 0});

        vector<bool> visitedNodes(
            totalNodes + 1,
            false
        );

        visitedNodes[1] = true;

        // Iterative DFS
        while (!dfsStack.empty()) {

            auto [currentNode, currentDepth] =
                dfsStack.top();

            dfsStack.pop();

            maximumDepth = max(
                maximumDepth,
                currentDepth
            );

            for (int neighborNode :
                 adjacencyList[currentNode]) {

                if (!visitedNodes[neighborNode]) {

                    visitedNodes[neighborNode] = true;

                    dfsStack.push(
                        {
                            neighborNode,
                            currentDepth + 1
                        }
                    );
                }
            }
        }

        // Fast modular exponentiation
        long long answer = 1;

        long long base = 2;

        int exponent = maximumDepth - 1;

        while (exponent > 0) {

            if (exponent & 1) {
                answer =
                    (answer * base) % MODULO;
            }

            base =
                (base * base) % MODULO;

            exponent >>= 1;
        }

        return (int)answer;
    }
};
